### ISSUE TITLE
BUG: Build failure with Xcode 12

### ISSUE THREAD

**Issue by jhpalmieri** [scipy/scipy#12860]
Xcode 12 for OS X was just released, and I can't build scipy with it. The command `pip3 install . -v --user` fails, ending with
```
  error: Command "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -I/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Tk.framework/Versions/8.5/Headers -DUSE_VENDOR_BLAS=1 -Iscipy/sparse/linalg/dsolve/SuperLU/SRC -I/private/var/folders/cp/n8wtqs490tq5psknff1hv9qr0000gn/T/pip-build-env-k4q2ai5g/overlay/lib/python3.8/site-packages/numpy/core/include -c scipy/sparse/linalg/dsolve/SuperLU/SRC/clacon2.c -o build/temp.macosx-10.15-x86_64-3.8/scipy/sparse/linalg/dsolve/SuperLU/SRC/clacon2.o -MMD -MF build/temp.macosx-10.15-x86_64-3.8/scipy/sparse/linalg/dsolve/SuperLU/SRC/clacon2.o.d" failed with exit status 1
  Building wheel for scipy (PEP 517): finished with status 'error'
  ERROR: Failed building wheel for scipy
Failed to build scipy
ERROR: Could not build wheels for scipy which use PEP 517 and cannot be installed directly
```

This may be the same issue as https://github.com/scipy/scipy/issues/12656 — at least the error at the end is the same — but that issue was closed while only providing the suggestion to use a different compiler. 

I am using Xcode 12, plus homebrew's Python 3.8, pip3, and gfortran. Unfortunately, I am really trying to figure out how to get this to build in Sage (sagemath.org), and using a gcc toolchain instead of clang would be, at the least, very inconvenient.

**Comment by dimpase** [comment]
To rectify, this happens with Xcode 12.2 beta.

**Comment by jhpalmieri** [comment]
It happens for me with Xcode 12.0 and OS X 10.15.6 — Catalina, not the prerelease of Big Sur. That is, it happens with the current Mac OS and Xcode, not beta or prerelease software.

**Comment by dimpase** [comment]
Yes, I can confirm this, too (i.e. both the stable Xcode 12.0 and Xcode 12.2.beta are not. Apple's versioning is far from perfect. Perhaps the easiest identifucation of the toolchain is from the C compiler (clang) version
```
Apple clang version 12.0.0 (clang-1200.0.32.4)
```
note that `xcode-select -v ` still outputs version 2373 - so it was not bumped.

A typical error comes from a missing `extern` declaration, such as 
```
scipy/sparse/linalg/dsolve/SuperLU/SRC/zlacon2.c:175:5: error: implicit declaration of function 'zcopy_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
      zcopy_(n, x, &c__1, v, &c__1);
      ^
```

I'll try to fix them all and make a PR

**Comment by dimpase** [comment]
Please see #12864 - backporting it to 1.5.2 requires more changes.

**Comment by adamjstewart** [comment]
> backporting it to 1.5.2 requires more changes.

What other changes are required? Will 1.5.3 be released soon? If not, a patch for 1.5.2 would be really helpful.

**Comment by dimpase** [comment]
> What other changes are required? Will 1.5.3 be released soon? 
I don't know - I am a very, very occasional contributor to scipy.

> If not, a patch for 1.5.2 would be really helpful.

we have a patch here: https://git.sagemath.org/sage.git/tree/build/pkgs/scipy/patches/extern_decls.patch?id=711fe05025795e44b84233e065d240859ccae5bd

**Comment by adamjstewart** [comment]
Thanks, that patch worked!

**Comment by dimpase** [comment]
Note that for backporting to 1.5.2 the qhull part ought to be taken from  #12864 - as it is cleaner and shorter.

**Comment by adamjstewart** [comment]
I'm trying to build SciPy 1.5.3 with Apple Clang 12.0.0 but I'm still seeing build issues:
```
...
extra options: '-std=c++14'
clang: /var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/tmpwvmdgti1/main.c
error: invalid argument '-std=c++14' not allowed with 'C'
C compiler: /Users/Adam/spack/lib/spack/env/clang/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -fPIC
...
compile options: '-c'
extra options: '-std=c++11'
clang: /var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/tmpj5quvw3n/main.c
error: invalid argument '-std=c++11' not allowed with 'C'
Could not detect c++ standard flag
C compiler: /Users/Adam/spack/lib/spack/env/clang/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -fPIC
...
scipy/sparse/linalg/dsolve/SuperLU/SRC/clacon2.c:175:5: error: implicit declaration of function 'ccopy_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    ccopy_(n, x, &c__1, v, &c__1);
    ^
...
scipy/sparse/linalg/dsolve/SuperLU/SRC/dmach.c:62:9: error: implicitly declaring library function 'strncmp' with type 'int (const char *, const char *, unsigned long)' [-Werror,-Wimplicit-function-declaration]
    if (strncmp(cmach, "E", 1)==0) {
        ^
...
scipy/sparse/linalg/dsolve/SuperLU/SRC/ilu_zdrop_row.c:196:3: error: implicit declaration of function 'dcopy_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                dcopy_(&len, dwork, &i_1, dwork2, &i_1);
                ^
...
scipy/sparse/linalg/dsolve/SuperLU/SRC/ilu_cdrop_row.c:196:3: error: implicit declaration of function 'scopy_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                scopy_(&len, swork, &i_1, swork2, &i_1);
                ^
...
scipy/sparse/linalg/dsolve/SuperLU/SRC/smach.c:62:9: error: implicitly declaring library function 'strncmp' with type 'int (const char *, const char *, unsigned long)' [-Werror,-Wimplicit-function-declaration]
    if (strncmp(cmach, "E", 1)==0) {
        ^
...
scipy/sparse/linalg/dsolve/SuperLU/SRC/zlacon2.c:175:5: error: implicit declaration of function 'zcopy_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    zcopy_(n, x, &c__1, v, &c__1);
    ^
...
scipy/sparse/linalg/dsolve/SuperLU/SRC/slacon2.c:160:16: error: implicit declaration of function 'idamax_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    isave[1] = idamax_(n, &x[0], &c__1);  /* j */
               ^
...
```
I'm going to try again with the previous patch.

**Comment by tylerjereddy** [comment]
@adamjstewart There's another proposed patch for 1.5.4 that includes a small change for SuperLU: https://github.com/scipy/scipy/pull/12957/files

@rlucas7 The patch Adam references above includes more changes than your PR though?

Seems to me like there's a bit of an issue with actual bug fixes getting mixed in with enhancements in PRs. Maybe I'll just have to check it carefully locally for 1.5.4...

**Comment by adamjstewart** [comment]
@tylerjereddy thanks, I'll try that out.

Is it possible to build and link against externally installed Qhull and SuperLU libraries? That might save you some of the trouble of maintaining your own patches, and might save build time since we have to build Qhull for matplotlib anyway.

Btw, are there any SciPy devs who would be interested in helping me maintain Spack's `py-scipy` package? You don't have to be a Spack expert, it just gives us someone to ping when a user reports a build issue or wants to modify the package recipe. For reference, here is what the recipe currently looks like: https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-scipy/package.py

**Comment by adamjstewart** [comment]
I tried the patch in #12957 but I still see the same build errors...

**Comment by rlucas7** [comment]
Yeah I had noticed this Discussikn thread when I opened PR @tylerjereddy it does look like there are more changes required. 

If that patch linked above contains fixes then we should use that one and I can close the 1-liner PR. 

-Lucas

**Comment by rgommers** [comment]
> Is it possible to build and link against externally installed Qhull and SuperLU libraries? 

No not possible. We have to vendor those libraries anyway, so also allowing building against external libs is (a) extra work to introduce, and (b) more variation in build configs, which in general isn't helpful.

**Comment by mfoglio** [comment]
How do I apply the patch to solve the issue?

**Comment by adamjstewart** [comment]
@mfoglio it looks something like:
```console
$ wget https://pypi.io/packages/source/s/scipy/scipy-1.5.3.tar.gz
$ tar xzf scipy-1.5.3.tar.gz 
$ cd scipy-1.5.3/
$ wget https://git.sagemath.org/sage.git/plain/build/pkgs/scipy/patches/extern_decls.patch?id=711fe05025795e44b84233e065d240859ccae5bd
$ patch -p1 <extern_decls.patch\?id\=711fe05025795e44b84233e065d240859ccae5bd 
patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/clacon2.c
patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/dmach.c
patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/ilu_cdrop_row.c
patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/ilu_zdrop_row.c
patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/slacon2.c
patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/smach.c
patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/sp_ienv.c
patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/zlacon2.c
patching file scipy/spatial/qhull_src/src/libqhull_r.h
```

**Comment by mfoglio** [comment]
I am using a virtual environment in poetry. Do you know if there is any old version of scipy that would not throw this error?

**Comment by mfoglio** [comment]
> @mfoglio it looks something like:
> 
> ```
> $ wget https://pypi.io/packages/source/s/scipy/scipy-1.5.3.tar.gz
> $ tar xzf scipy-1.5.3.tar.gz 
> $ cd scipy-1.5.3/
> $ wget https://git.sagemath.org/sage.git/plain/build/pkgs/scipy/patches/extern_decls.patch?id=711fe05025795e44b84233e065d240859ccae5bd
> $ patch -p1 <extern_decls.patch\?id\=711fe05025795e44b84233e065d240859ccae5bd 
> patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/clacon2.c
> patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/dmach.c
> patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/ilu_cdrop_row.c
> patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/ilu_zdrop_row.c
> patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/slacon2.c
> patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/smach.c
> patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/sp_ienv.c
> patching file scipy/sparse/linalg/dsolve/SuperLU/SRC/zlacon2.c
> patching file scipy/spatial/qhull_src/src/libqhull_r.h
> ```


I tried this but then the installation https://docs.scipy.org/doc/scipy/reference/building/macosx.html fails with a really long error

```
....
13 errors generated.
error: Command "clang++ -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -I/Users/matteo/.pyenv/versions/3.7.7/lib/python3.7/site-packages/numpy/core/include -c scipy/optimize/rectangular_lsap/rectangular_lsap.cpp -o build/temp.macosx-10.15-x86_64-3.7/scipy/optimize/rectangular_lsap/rectangular_lsap.o -MMD -MF build/temp.macosx-10.15-x86_64-3.7/scipy/optimize/rectangular_lsap/rectangular_lsap.o.d -mmacosx-version-min=10.9" failed with exit status 1
```

I can't find a way to recreate my poetry environment.

**Comment by tylerjereddy** [comment]
The TODOs here seem to be:

1) Determine what patch is needed for `master` branch. I was, for some reason,  under the impression this was fixed for `master` but not the `1.5.x` maintenance branch. gh-12937 suggests that `master` has issues as well unless the CI there is using yet another version of XCode or something? (the CI failed there even before my patching attempt). Conversely, when I've tried local builds of `master` with XCode 12 they seem to work just fine.

2) We seem pretty close on a patch that will fix the issue for `1.5.x` based on the feedback from i.e., @adamjstewart above in gh-12957. What I might do there is turn on the same MacOS Python 3.9 CI that is failing in gh-12937 and if that passes we can maybe move forward with 1.5.x in that regard at least. (We may be able to remove the qhull shims though, since a simpler one should have been backported). This patch actually does show the fail before/fix after applying when I test locally with XCode 12.
